packages/kbn-i18n/src/core.js

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import IntlMessageFormat from 'intl-messageformat';
+import memoizeIntlConstructor from 'intl-format-cache';
+
+const EN_LOCALE = 'en';
+
+const getMessageById = (messages, id) =>
+  id.split('.').reduce((val, key) => (val ? val[key] : null), messages);
+
+const hasValues = values => Object.keys(values).length > 0;
+
+const addLocaleData = localeData => {
+  if (localeData && localeData.locale) {
+    IntlMessageFormat.__addLocaleData(localeData);
+  }
+};
+
+const normalizeLocale = (locale = '') => locale.toLowerCase().replace('_', '-');
+
+const showError = error => {
+  if (process.env.NODE_ENV !== 'production') {
+    console.error(error);
+  }
+};
+
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+
+export default class I18n {
+  constructor(messages = {}, locale = messages.locale || EN_LOCALE) {
+    this.currentLocale = normalizeLocale(locale);
+    this.messages = { [this.currentLocale]: messages };
+    this.defaultLocale = EN_LOCALE;
+    this.formats = {};
+
+    if (messages.localeData) {
+      addLocaleData(messages.localeData);
+    }
+  }
+
+  addMessages(messages = {}, locale = messages.locale || EN_LOCALE) {
+    const normalizedLocale = normalizeLocale(locale);
+
+    this.messages[normalizedLocale] = {
+      ...this.messages[normalizedLocale],
+      ...messages,
+    };
+
+    if (messages.localeData) {
+      addLocaleData(messages.localeData);
+    }
+  }
+
+  getMessages() {
+    return this.messages[this.currentLocale];
+  }
+
+  setLocale(locale) {
+    this.currentLocale = normalizeLocale(locale);
+  }
+
+  getLocale() {
+    return this.currentLocale;
+  }
+
+  setDefaultLocale(locale) {
+    this.defaultLocale = normalizeLocale(locale);
```

Comment:
__nit:__ it'd probably make sense to do some basic argument type checking in this and other methods within this class. E.g.  throw if `locale` is not non-empty string or `messages` arg isn't defined and not an object etc. Just very light checks, nothing advanced.